### ISSUE TITLE
[templates] Source/target versions in bnd not set to match the version of java selected

### ISSUE THREAD

**Issue by bjhargrave** [bndtools/bndtools#1203]
The new templates are a mess. I tried to create a simple DS project in a new workspace and received a number of errors rather than a working project.

1.) I chose the project name "ds-1.3" which is a legit project name but not a valid package name. The template did not notice and fix/warn on this.

2.) The test class was not put in a test source folder (the test package was place in the base folder of the project.)

3.) No test entry was placed in the .classpath file.

4.) The junit and hamcrest entries placed in the bnd file were not valid for the JPM repo I chose when creating the cnf project.

5.) The source/target versions in bnd were not set to match the version of java selected in the new project wizard. So I get 3 warnings about the source/target/JRE mismatch to be fixed.

Also, where are the enroute templates from 3.0?

**Comment by pkriens** [comment]
I have already talked to Neil because I found the new templates wanting. 

First, the disappearance of 3.0 enRoute templates are gone is actually quite serious. Trying to port them I found missing features. I think we should always do our utmost to NOT remove functionality unless we mutually agree. Even though this is alpha, it is extremely inconvenient if you miss features.

* The the templating misses test folder support, and 
* It is impossible to use code to do templating as was the case before, and
* Using the default $ for the templating is awkward because there are so many macros that need escaping.

I suggest we revert the templates to 3.0 asap. I've made some changes to the current model that I think we need to standardize like being able to use code.

**Comment by pkriens** [comment]
@njbartlett wdyt?

**Comment by njbartlett** [comment]
If there are gaps or bugs in the new template system then let's fix those. I'd be very very unhappy about going back on this.

Peter, you're arguing against progress. My expectation was that you would port the enRoute templates before the 3.1 release. I'm willing to help but since enRoute is your baby I think this has to be something you drive. I always expected the process to expose bugs that needed to be fixed... but the new templates have been in for several weeks, so I'm kind of surprised to be hearing about this now.

**Comment by bjhargrave** [comment]
But you can't just blow away function and then expect someone else to figure out how to make it work again. At least not without prior agreement.

You are hearing about it now because we are trying to use the templates for EclipseCon sessions and cannot. So we must use 3.0 for EclipseCon.

**Comment by njbartlett** [comment]
That's an unfair representation of what happened. I told Peter about what I was doing and why, I did a *lot* of work on the core functionality, but I didn't know what else needed to be there for enRoute. I expected to work with Peter but I only heard from him one week before EclipseCon!

Using 3.0 for EclipseCon sounds like the best plan all around, doesn't it?

**Comment by pkriens** [comment]
We talked but it was never clear to me that you were going to remove the old functionality. I expected to be able to migrate them over time. I actually still find it hard to see that you don't see how painful this is for me and others.

And we have the problem that the functionality of the new system is much less than the old system.

**Comment by pkriens** [comment]
Can't you revert to the old system and at the same time support the new templates?

**Comment by njbartlett** [comment]
Why would we want to release 3.1 with two different template systems?

You're right, I still don't see how painful this is for you! 3.0 exists. It's stable. I plan to use it for my demos at EclipseCon and I expected you would do... it would be crazy to use the master branch for a tutorial. If master is not the place to be building major new functionality then where the hell is?

For my part, I find this conversation deeply frustrating, and bordering on offensive.

If the new template system is truly unfixable, then I will certainly revert it, but I'm not persuaded of that yet. If it is fixable, then I'll fix it. We can start by raising individual bugs for the problems BJ enumerated at the top of this thread and assigning them to me.

**Comment by pkriens** [comment]
I am not planning to use 3.1 for the CE but I hate it that I do not have an escape when we find a bug in 3.0. Remember last year when we did an emergency release for you for exactly such a reason. The removal of the enRoute templates put a significant amount of work on me to make 3.1 on parity with 3.0.

Anyway, lets see if we can move this forward. If you're enumerating bugs please take a look at the enRoute templates and analyze the way they work. On the top of my head, I used the following functionality that you all seem to have removed from the new system:

* Optional test folders (no we seem to never have a test folder)
* Set information in the model before presented to the user
* Numerous extra variables. In enRoute's case based on variations of the project name
* Regular expression matching for replacement
* Templates must be able to build in a bndtools 'as is' (the escaping of the $ is a big problem)
* Need to have a separate set of variables for the file paths and the content
* Allow the use of code to decide what/how
* Be part of a bndtools feature if so desired (we now only have 1 hardcoded built in template)
* Run a build at the end of the copying

Basically the API we used to have covered this all, not sure why that had to be replaced? The string templating could then have been a convenience function, expressed as a base class for example.

A new requirement that I would have to loved to see is a question form with template specific questions. I guess this is what they do in Amdatu Bootstrap.

This all does not mean I don't like the new capability to load them from a repository, that is a great idea and I think the way forward.

**Comment by njbartlett** [comment]
First problem in BJ's list is addressed in 3aff73680b904bddc81e52a7d1e0ae10da8cb6ab

**Comment by njbartlett** [comment]
Picking out a few of these:

>    Optional test folders (no we seem to never have a test folder)
>    Set information in the model before presented to the user

Clarification please?

>    Numerous extra variables. In enRoute's case based on variations of the project name

Rather than having logic producing entirely different output depending on the name of the project, why not have separate templates for each output type?

>    Regular expression matching for replacement
>    Templates must be able to build in a bndtools 'as is' (the escaping of the $ is a big problem)

Already addressed (and we discussed this!) with the ability to use a different delimiter character (see `leftDelim` and `rightDelim` in `_template.properties`.

>    Need to have a separate set of variables for the file paths and the content
>    Allow the use of code to decide what/how

This seems problematic if templates are located in the repository... we're essentially going to download and immediately run untrusted code. I'm not convinced of the requirement for templates to contain executable code.

>    Be part of a bndtools feature if so desired (we now only have 1 hardcoded built in template)
>    Run a build at the end of the copying

This should be happening already. Obviously you need to make sure Eclipse has got automatic builds turned on.

**Comment by njbartlett** [comment]
Test folder issue addressed in 51979dd8220a2e0c0b34625eefac8ea8e51a1902.

We now detect whether the template references the `testSrcDir` parameter, and only add the test source dir if it does.

The same mechanism could be extended to check for additional template parameters and collect values for them from the UI.

**Comment by pkriens** [comment]
**Test folder:** why not like the old days when I could just answer a method call if a test folder is included? Referring a parameter sounds awfully indirect? I want to be able to calculate this, or worst case specify that in a configuration file. (Though I prefer code.)

**Model:** In the old model we could specify some of the parameters shown to the user in the wizard (for example proposed layout of the src folders). This is lost now since the templates are now passive. It is not that crucial for me but it was rather nice to be able to talk to the user and set the defaults. As said, in the future we should be able to ask questions to the user and use the answers in the setup.

**Type of template:** Yes, I plan to have different template names in the new model. Would be nice though if I could have one of the templates selected based on the name. However, that is not the issue. I created lots of names based on the project name. I.e. instead of Example.java I call it SpeakerImpl.java if the project is like *.speaker.provider. I also calculate upper case versions for headers, PIDs, etc. This works fairly well. (Did you ever try an enRoute template?)

**The _ and $ delimiter:** Yes I saw this, but the default of $ requires work for the common case that you want your template buildable by bndtools before expansion. With the enRoute templates I started out by making them as just a resource (like you did in the current templates) but that quickly drove me crazy because you don't get any help from bnd/Bndtools to prevent errors in the setup. I think therefore that the default should be a character like _ instead of something that always conflicts with bnd.

**Run code:** I think the String templating you use now is just not sufficient for a lot of use cases, it is excluding too many possibilities. I think String templating is great for very simple templates. However, I think you need an escape since you never know what a template designer must be able to do. I love convenience of a simple model but I really dislike it if they become straight jackets. In such cases I tend to make my preferred model very easy, thereby enticing the designer to use it, but I always try to offer an escape because I do not know their actual problems. 

Yes there is a security issue. So we could require that code to be installed as a feature in Eclipse but I think we should be able to load it from the repository as well (worst care give a warning the first time or mark the templates in the select list). And we're already running code from repositories all the time. Anytime you launch your program you run code from the repositories.

**Build at end**: Since I could not run the enRoute templates in the current setup I could not verify if this works. However, in the 3.0 setup I had to add a build at the end because I was overwriting the .project and .classpath files. Very often I ended with a project that had errors and needed to be cleaned. Doing a build at the end cleaned this up. Might not happen in this case, but just wanted to make sure it is in there.

I still think there is a gap in architecture where you are now and they features of 3.0 on which the enRoute templates depend. It also sounds like you're not happy to support some of the old features like support for running code. I am ok using string templating as a preferred way to make simple templates but I am not happy forcing all templates to a severely constrained model that does not allow the things we could do before. (Notice how Amdatu made this a completely separate product with their bootstrap.) 

It seems seem that we have different ambitions in what templates should be able to do, which is a fair and we need to work this out. However, this tends to take time. Unfortunately, during that time I am screwed for using 3.1 and have no backup if we need to make an emergency fix. I do not think that that is fair to me, nor to enRoute users. So I really think we need the old functionality back in the mean time. I am ok to have both the old and new templating (extension points) there at the same time for now if you do not want to support running code from the repositories.

Last last Friday I tried to port the enRoute templates to the string templating model and started to make changes to the templating library but had to give up because it was too big a change. More than willing to share that code to help you out on a separate branch.

**Comment by bjhargrave** [comment]
> 5.) The source/target versions in bnd were not set to match the version of java selected in the new project wizard. So I get 3 warnings about the source/target/JRE mismatch to be fixed.

This issue still exists when I create a project. My eclipse has 1.8 as the default JDK which is the default used when creating the project. But the cnf wizard just takes the bnd javac defaults of 1.5

```
Description	Resource	Path	Location	Type
Eclipse: javac.source inconsistency between bnd & Eclipse. bnd is 1.5 and Eclipse is 1.8	bnd.bnd	/test2	line 1	Bndtools Path Problem Marker
Eclipse: javac.target inconsistency between bnd & Eclipse. bnd is 1.5 and Eclipse is 1.8	bnd.bnd	/test2	line 1	Bndtools Path Problem Marker
Eclipse: The .JRE container is set to JavaSE-1.8 but bnd is compiling against 1.5	.classpath	/test2	line 1	Bndtools Path Problem Marker
```

**Comment by bjhargrave** [comment]
> 4.) The junit and hamcrest entries placed in the bnd file were not valid for the JPM repo I chose when creating the cnf project.

This problem still exists as well. Using the bundle-hub, the `-testpath` entries are ok. But if you create your cnf using the JPM repo, the created `-testpath` does not resolve to any bundles.

```
Description	Resource	Path	Location	Type
Cannot find /error/hamcrest-core;version=1.3 Not found in [bnd-cache, Local, JpmRepository [writable=false, name=JPM4J, digest=83F7A05D6C75D70206207BC12B576AFCFAA94E14], Release]	bnd.bnd	/dstest	line 6	Bndtools Path Problem Marker
Cannot find /error/junit;version=4 Not found in [bnd-cache, Local, JpmRepository [writable=false, name=JPM4J, digest=83F7A05D6C75D70206207BC12B576AFCFAA94E14], Release]	bnd.bnd	/dstest	line 6	Bndtools Path Problem Marker
hamcrest-core;version=1.3 Not found in [bnd-cache, Local, JpmRepository [writable=false, name=JPM4J, digest=83F7A05D6C75D70206207BC12B576AFCFAA94E14], Release]	bnd.bnd	/dstest	line 6	Bndtools Path Problem Marker
junit;version=4 Not found in [bnd-cache, Local, JpmRepository [writable=false, name=JPM4J, digest=83F7A05D6C75D70206207BC12B576AFCFAA94E14], Release]	bnd.bnd	/dstest	line 6	Bndtools Path Problem Marker
```

**Comment by njbartlett** [comment]
I lost track of where we are with this bug, and it still has milestone 3.1. Can we close it, or does it need to move to 3.2?

**Comment by bjhargrave** [comment]
I think [item 5](https://github.com/bndtools/bndtools/issues/1203#issuecomment-165168827) still needs to be fixed for 3.2.

**Comment by njbartlett** [comment]
Ok moved to 3.2 milestone.

**Comment by njbartlett** [comment]
Updated title to reflect narrowed scope

**Comment by njbartlett** [comment]
The source/target versions is basically addressed by commit 5b8fcb58a05fd9f935b1f9f5f9ad239758f6af1d.

Closing... BJ please let me know whether this issue is solved to your satisfaction.
